### PR TITLE
[PT-189] Percentage calculation

### PR DESCRIPTION
**Pull request by 0xbuilder0**
# [PT-189](https://0xbase.atlassian.net/browse/PT-189) - Percentage calculation

### Description

- Increase the precision size for TotalPrice
- External API error handling with response status code and handle in code and change the interface for fetch data

[PT-189]: https://0xbase.atlassian.net/browse/PT-189?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Comment by 0xbuilder0** [comment]
This issue is due the the precision value so for example when we calculate the total value of asset which is UnitPrice * Quantity  which result the TotalPrice . So currently roundoff the  Asset TotalPrice  as well as the Portfolio TotalPrice  are roundoff up-to 3 precision . So now just imagine the UnitPrice * Quantity  give the result e.g 0.000005744  and due to 3 precision  give it end up to 0  0.000005744 which result the 0 value... Now can with make the precision  up to 7 or 10... need to decide.

File: providers/coingecko/coingecko.go [hunk 1]
```diff
@@ -12,16 +12,15 @@ type (
 	CoingeckoAPI struct{}
 )
 
-func (c *CoingeckoAPI) FetchData(cryptoID, currency string) ([]byte, error) {
+func (c *CoingeckoAPI) FetchData(cryptoID, currency string) ([]byte, int, error) {
 	url := fmt.Sprintf("https://api.coingecko.com/api/v3/simple/price?ids=%s&vs_currencies=%s", cryptoID, currency)
```
Comment (fuddin95) [inline providers/coingecko/coingecko.go:16]: lets keep constant in a separate file `https://api.coingecko.com/api/v3/simple/price?ids=%s&vs_currencies=%s`

File: providers/debank/debank.go [hunk 2]
```diff
@@ -19,38 +20,46 @@ type (
...
 	headers := map[string]string{
 		"Accept":    "application/json",
 		"AccessKey": configs.EnvConfigVars.GetDebankAccessKeyHeader(),
 	}
 
 	resp := EvmDebankTotalBalanceApiResponse{}
-	if err := d.fetch("https://pro-openapi.debank.com/v1/user/total_balance?id="+address, headers, &resp); err != nil {
-		return nil, err
+
+	statusCode, err := d.fetch("https://pro-openapi.debank.com/v1/user/total_balance?id="+address, headers, &resp)
```
Comment (fuddin95) [inline providers/debank/debank.go:31]: same here

File: providers/debank/debank.go [hunk 3]
```diff
@@ -19,38 +20,46 @@ type (
...
-	if err := d.fetch("https://pro-openapi.debank.com/v1/user/all_token_list?id="+address, headers, &resp.TokensList); err != nil {
-		return nil, err
+	statusCode, err = d.fetch("https://pro-openapi.debank.com/v1/user/all_token_list?id="+address, headers, &resp.TokensList)
+	if err != nil {
+		return nil, statusCode, err
 	}
 
-	if err := d.fetch("https://pro-openapi.debank.com/v1/user/all_nft_list?id="+address, headers, &resp.NFTList); err != nil {
-		return nil, err
+	statusCode, err = d.fetch("https://pro-openapi.debank.com/v1/user/all_nft_list?id="+address, headers, &resp.NFTList)
```
Comment (fuddin95) [inline providers/debank/debank.go:41]: same here

File: providers/debank/debank.go [hunk 4]
```diff
@@ -19,38 +20,46 @@ type (
...
-		return nil, err
+
+	statusCode, err := d.fetch("https://pro-openapi.debank.com/v1/user/total_balance?id="+address, headers, &resp)
+	if err != nil {
+		return nil, statusCode, err
 	}
 
-	if err := d.fetch("https://pro-openapi.debank.com/v1/user/all_token_list?id="+address, headers, &resp.TokensList); err != nil {
-		return nil, err
+	statusCode, err = d.fetch("https://pro-openapi.debank.com/v1/user/all_token_list?id="+address, headers, &resp.TokensList)
```
Comment (fuddin95) [inline providers/debank/debank.go:36]: same here

File: providers/moralis/moralis.go [hunk 5]
```diff
@@ -19,17 +19,17 @@ type (
 	MoralisAPI struct{}
 )
 
-func (s *MoralisAPI) FetchData(address string) ([]byte, error) {
+func (s *MoralisAPI) FetchData(address string) ([]byte, int, error) {
 	url := "https://solana-gateway.moralis.io/account/mainnet/" + address + "/portfolio"
```
Comment (fuddin95) [inline providers/moralis/moralis.go:23]: same here

**Review by fuddin95 — APPROVED** [review]

File: providers/coingecko/coingecko.go (hunk 1 above)
Comment (0xbuilder0) [reply providers/coingecko/coingecko.go]: fixed

File: providers/debank/debank.go (hunk 2 above)
Comment (0xbuilder0) [reply providers/debank/debank.go]: fixed

File: providers/debank/debank.go (hunk 3 above)
Comment (0xbuilder0) [reply providers/debank/debank.go]: fixed

File: providers/debank/debank.go (hunk 4 above)
Comment (0xbuilder0) [reply providers/debank/debank.go]: fixed

File: providers/moralis/moralis.go (hunk 5 above)
Comment (0xbuilder0) [reply providers/moralis/moralis.go]: fixed

**Pull request merged**